File: 12a.py
#! /usr/bin/python3
import sys

caves = {}
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        (From, To) = line.rstrip("\n").split('-')
        if From in caves:
            caves[From].add(To)
        else:
            caves[From] = {To}
        if To in caves:
            caves[To].add(From)
        else:
            caves[To] = {From}

paths = [['start']]
while True:
    extendable = list(reversed(list(filter(lambda i:paths[i][-1] != 'end',
                                           range(len(paths))))))
    if not extendable:
        break

    for ie in extendable:
        e = paths.pop(ie)

        for nxt in caves[ e[-1] ]:
            new = e.copy()
            new.append(nxt)
            if nxt.lower() == nxt and nxt in e:
                continue
            paths.append(new)

print(len(paths))

File: 12a.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my %caves;
while (<>) {
    chomp;
    my ($from, $to) = split /-/;
    undef $caves{$from}{$to};
    undef $caves{$to}{$from};
}

my @paths = (['start']);
while (my @extendable = grep $paths[$_]->[-1] ne 'end', 0 .. $#paths) {
    for my $ie (reverse @extendable) {
        my $e = splice @paths, $ie, 1;
        for my $next (keys %{ $caves{ $e->[-1] } }) {
            push @paths, [@$e, $next]
                unless $next =~ /^[[:lower:]]/o
                && grep $_ eq $next, @$e;
        }
    }
}
say scalar @paths;

__END__
start-A
start-b
A-c
A-b
b-d
A-end
b-end
